### PR TITLE
Validator: Support SPV_NV_raw_access_chains

### PR DESCRIPTION
**Pull request by rlocatti-nv**
Add support for https://github.com/KhronosGroup/SPIRV-Registry/pull/238 .

### CODE REVIEW

File: source/val/validate_memory.cpp [hunk 1]
```diff
@@ -1430,6 +1430,85 @@ spv_result_t ValidateAccessChain(ValidationState_t& _,
   return SPV_SUCCESS;
 }
 
+spv_result_t ValidateRawAccessChain(ValidationState_t& _,
+                                    const Instruction* inst) {
+  std::string instr_name = "Op" + std::string(spvOpcodeString(inst->opcode()));
+
+  if (!_.features().raw_access_chains) {
```
Comment (alan-baker) [inline source/val/validate_memory.cpp:1437]: This requirement is already communicated via the grammar.

File: source/val/validate_memory.cpp [hunk 2]
```diff
@@ -1430,6 +1430,85 @@ spv_result_t ValidateAccessChain(ValidationState_t& _,
...
+    return _.diag(SPV_ERROR_INVALID_ID, inst)
+           << "The Result Type of " << instr_name << " <id> "
+           << _.getIdName(inst->id()) << " must be OpTypePointer. Found Op"
+           << spvOpcodeString(result_type->opcode())
+           << '.';
+  }
+
+  // The pointed storage class must be valid.
+  const auto storage_class = result_type->GetOperandAs<spv::StorageClass>(1);
+  if (storage_class != spv::StorageClass::Workgroup &&
```
Comment (alan-baker) [inline source/val/validate_memory.cpp:1455]: Workgroup is not listed as a valid storage class in the [spec](https://htmlpreview.github.io/?https://github.com/KhronosGroup/SPIRV-Registry/blob/main/extensions/NV/SPV_NV_raw_access_chains.html). Is that a spec bug or tooling bug?

File: source/val/validate_memory.cpp [hunk 3]
```diff
@@ -1430,6 +1430,85 @@ spv_result_t ValidateAccessChain(ValidationState_t& _,
...
+      result_type_pointee->opcode() == spv::Op::OpTypeMatrix ||
+      result_type_pointee->opcode() == spv::Op::OpTypeStruct) {
+    return _.diag(SPV_ERROR_INVALID_ID, inst)
+           << "The Result Type of " << instr_name << " <id> "
+           << _.getIdName(inst->id()) << " must not point to "
+           "OpTypeArray, OpTypeMatrix, or OpTypeStruct.";
+  }
+
+  const auto stride = _.FindDef(inst->GetOperandAs<uint32_t>(3));
+  if (stride->opcode() != spv::Op::OpConstant) {
```
Comment (alan-baker) [inline source/val/validate_memory.cpp:1477]: There is no type checking of stride, index, or offset. Is that intentional?

File: source/val/validation_state.h [hunk 4]
```diff
@@ -71,6 +71,9 @@ class ValidationState_t {
                                          // and its values to be used without
                                          // requiring any capability
 
+    // Allow using raw access chains through the RawAccessChainsEXT capability
+    bool raw_access_chains = false;
```
Comment (alan-baker) [inline source/val/validation_state.h:75]: I think this is unnecessary. If feature is checked for it should be done via `HasCapability(spv::Capability::RawAccessChainsNV)`.

**Review by alan-baker — CHANGES_REQUESTED** [review]
Please also update the DEPS file to point to the latest SPIRV-Headers.

**Review by alan-baker — COMMENTED** [review]
It would also be useful to have unit tests for the new functionality.

File: source/val/validate_memory.cpp (hunk 2 above)
Comment (rlocatti-nv) [reply source/val/validate_memory.cpp]: It was a bug here. Removed it.

File: source/val/validate_memory.cpp (hunk 3 above)
Comment (rlocatti-nv) [reply source/val/validate_memory.cpp:1468]: Added type checking. It was not intentional.

**Comment by alan-baker** [comment]
You'll need to update DEPS to include the SPIRV-Headers that includes raw access chains for the bots to work correctly.

**Comment by alan-baker** [comment]
Now the bots are complaining about a redefinition of the operand enums. Please rebase on main.

File: test/val/val_extension_spv_nv_raw_access_chains.cpp [hunk 5]
```diff
@@ -0,0 +1,491 @@
...
+
+    %main = OpFunction %void None %mainFunctionType
+    %label = OpLabel
+    %rawChain = OpRawAccessChainNV %intPtr %ssbo %int_16 %int_0 %int_0 RobustnessPerComponentNV
+    %unused = OpLoad %int %rawChain
+    OpReturn
+    OpFunctionEnd
+)";
+  CompileSuccessfully(str.c_str());
+  EXPECT_EQ(SPV_ERROR_INVALID_CAPABILITY, ValidateInstructions());
```
Comment (alan-baker) [inline test/val/val_extension_spv_nv_raw_access_chains.cpp:104]: When testing error cases, the style is generally to have an additional expectation along the lines of:
```
EXPECT_THAT(getDiagnosticString(),
                          HasSubstr("error string or relevant substring to confirm right error is triggered"));
```

**Review by alan-baker — CHANGES_REQUESTED** [review]
Also formatting needs updated.

**Comment by HansKristian-Work** [comment]
Is there any ETA on when we get some forward progress on this PR? I'm waiting for this PR to get merged.

**Review by alan-baker — APPROVED** [review]

**Pull request merged**